### ISSUE TITLE
Add higher level primitives for FTP cwd/pwd

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2228]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2228|
|Type|enhancement|
|Created|2006-11-13 09:17:25Z|

We should have more useful methods for cwd/pwd, with clearer names, and that don't need to use parsePWDResponse. I propose the names setDirectory/getDirectory.

Attachments:

* [ftp_2228.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/efc/efc617ea7ae446105d8e4a6c58aac2148947dba5/3c5c936b8a40691710c946905e22f74a3b9d2d5a.diff) (12480 bytes) - added by therve on 2006-12-23 08:58:38Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2228 2228
type__enhancement enhancement
reporter__therve therve
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1163409445000000 1163409445000000
changetime__1166984133000000 1166984133000000
version__None None
owner__ 
cc__spiv
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

See [#8053](https://github.com/twisted/twisted/issues/8053) for initial discussion.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|spiv commented|
|-|-|

I'm in favour of the basic idea of useful higher-level functions (with more English-like names) on top of "cwd" and "pwd".  One comment: "changeDirectory" would be clearer than "setDirectory", I think.  Most people are familiar with the concept from the "cd" shell command, and the `os.chdir` function.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

I made a first try. I tried to be strict in parsing result, but was not so inspired in tests.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @therve|
|-|-|

Docstrings should look
```python
"""
Like this
"""
```
not
```python
"""Like this
"""
```

Docstrings should say what kinds of failures will be generated.

Test method names should not start with a capital letter after the ```test_``` prefix.

Correct epytext is ```@return```, not ```@returns```, despite the fact that the output says "returns".

Please submit a patch that fixes these issues.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

DOne, I fixed the @returns in all ftp.py and doctring format in FTPClient for consistency.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_6"></a>|@jerub commented|
|-|-|

Excellent, thanks for updating all those docstrings to look the same all through ftp.py.

I think this is in good enough state to merge. Please make a clear note in your commit message that you're adding two methods to the api.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [19061]) Apply FTPClient pwd/cwd improvement patch

Author: therve
Reviewer: jerub
Fixes [#2228](https://github.com/twisted/twisted/issues/2228)

Add two new methods to FTPClient, getDirectory and changeDirectory, which provide a
higher-level interface to the existing pwd and cwd methods and removing the need for
application-level parsing of FTP protocol bytes in order to interpret the results of
those commands.

Also clean up various docstring formatting and epytext markup.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
